### DISCUSSION TITLE
Jinja2 and Django-unicorn

### DISCUSSION THREAD

**Discussion by bschollnick** [django-commons/django-unicorn#495] [Q&A]
Folks,

I'm trying to get Django-unicorn to play nicely with Jinja2 templates (via Jinja2 and Django-Jinja.

This isn't so crazy as it sounds...  But requires a couple of tags to be added to the Jinja2 environment.

e.g.

`
from django.templatetags.static import static
from django.urls import reverse
from django_icons.templatetags.icons import icon_tag
from jinja2 import Environment
from django_unicorn.templatetags.unicorn import unicorn_scripts, unicorn

def environment(**options):
    env = Environment(**options)
    env.globals.update({
        'static': static,
        'url': reverse,
        'icon':icon_tag,
        'unicorn':unicorn,
        'unicorn_scripts':unicorn_scripts,
    })
    return env
`

when used in a jinja2 template, e.g.

`
<!DOCTYPE html>
{{ unicorn }}
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge"/>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma@0.8.2/css/bulma.min.css">
    <script src="https://use.fontawesome.com/releases/v5.3.1/js/all.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery.lazy/1.7.9/jquery.lazy.min.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery.lazy/1.7.9/jquery.lazy.plugins.min.js"></script>
    <title>{% block title %} {{ gallery_name }} {% endblock %}</title>
</head>
<body>
`

I'm getting an error stating that it requires parameters.
Request Method: | GET
-- | --
http://nerv.local:8888/albums/-these/2010-10-10?page=1&sort=0
4.1.4
TypeError
unicorn() missing 2 required positional arguments: 'parser' and 'token'

Which makes sense due to the parameters in the function...

But according to the documentation, unicorn doesn't have any parameters?

-----------
4. Add {% unicorn_scripts %} into the Django HTML template and make sure there is a {% csrf_token %} in the template as well.

{% load unicorn %}
<html>
  <head>
    {% unicorn_scripts %}
  </head>
  <body>
    {% csrf_token %}
  </body>
</html>
-----------

So what am I missing here?  Did I misunderstand something?

**Comment by adamghill** [comment]
Hmm, this is not something I've ever tried to do. :)

To answer your question, `parser` and `token` arguments are used for the [`unicorn` function](https://github.com/adamghill/django-unicorn/blob/main/django_unicorn/templatetags/unicorn.py#L40) because it's a custom templatetag that jumps through a bunch of hoops to parse the args, kwargs, render the component (similar to how the `include` templatetag works if I remember correctly), etc. That all happens in [`render`](https://github.com/adamghill/django-unicorn/blob/main/django_unicorn/templatetags/unicorn.py#L69). For more details on how this works, check out [Writing the compilation function](https://docs.djangoproject.com/en/4.1/howto/custom-template-tags/#writing-the-compilation-function) in the docs.

I look forward to seeing if this is possible and what you come up with!
